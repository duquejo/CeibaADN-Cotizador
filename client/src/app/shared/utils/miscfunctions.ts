import moment from 'moment';
import { ISelectionOptions } from '../../feature/Admin/models/FormSelector';

export const obtenerCalendarioActivo = (
    calendarioActivo: string | number | null = null, 
    todosCalendarios: ISelectionOptions[] ) => {

    const estaActivoEnCalendarios = todosCalendarios.some( 
        ( calendario: ISelectionOptions ) => {
        return Number( calendario.value ) === Number( calendarioActivo );
    });

    if( calendarioActivo === "" || ! estaActivoEnCalendarios ) {
      if( todosCalendarios.length > 0 && todosCalendarios[0].value ) {
        return Number( todosCalendarios[0].value );
      } else {
        return null;
      }
    }
    return Number( calendarioActivo );
}

export const extraerIdsDeArray = ( needle: Array<any>, haystack: Array<any> ) => {
  
  if( ! needle.length ) {
    return [];
  }

  const needleIds = needle.map( ( elemento: any ) => {
    if( ! elemento.hasOwnProperty( 'id' ) ){
      return elemento;
    }
    return elemento.id;
  } );

  return haystack.filter( ( elemento: any ) => {
    return needleIds.indexOf( elemento!.id ) !== -1
  });
}

export const cotizacionTemplate = ( cotizacion: any, allInfo = false ) => {
  return `<table class="resultados__table">
            <caption>Información general</caption>
            <thead>
              <tr>
                <th>Identificador</th>
                <th>Código</th>
                <th>Fecha de inicio</th>
                <th>Fecha de finalización</th>
                <th># de personas</th>
                <th>Total grupo</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>${ cotizacion.id }</td>
                <td>${ cotizacion.codigo }</td>
                <td>${ moment( cotizacion.fechaInicio ).format('YYYY-MM-DD') }</td>
                <td>${ moment( cotizacion.fechaFin ).format('YYYY-MM-DD') }</td>
                <td>${ cotizacion.personas }</td>
                <td>$ ${ cotizacion.total }</td>
              </tr>
            </tbody>
          </table>
          <table class="resultados__table">
            <caption>Información del centro vacacional</caption>
            <thead>
              <tr>
                ${ allInfo ? `<th>Identificador</th>` : `` }
                <th>Nombre</th>
                <th>Descripción</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                ${ allInfo ? `<td>${ cotizacion.centroVacacional?.codigo }</td>` : `` }
                <td>${ cotizacion.centroVacacional?.nombre }</td>
                <td>${ cotizacion.centroVacacional?.descripcion }</td>
              </tr>
            </tbody>
            <table class="resultados__table">
              <caption>Información de la categoría de usuario</caption>
              <thead>
                <tr>
                  ${ allInfo ? `<th>Identificador</th>` : `` }
                  <th>Nombre</th>
                  <th>Descripción</th>
                  <th>Valor de día temporada alta</th>
                  <th>Valor de día temporada baja</th>
                </tr>
              </thead>
              <tbody>
                <tr>
                  ${ allInfo ? `<td>${ cotizacion.categoriaUsuarios.id }</td>` : `` }
                  <td>${ cotizacion.categoriaUsuarios.nombre }</td>
                  <td>${ cotizacion.categoriaUsuarios.descripcion }</td>
                  <td>$ ${ cotizacion.categoriaUsuarios.valorAlta }</td>
                  <td>$ ${ cotizacion.categoriaUsuarios.valorBaja }</td>
                </tr>
              </tbody>                
            </table>
          </table>`;
}