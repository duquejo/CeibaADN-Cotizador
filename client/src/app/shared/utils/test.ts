export const setTextEvent = (identifier: string, text: string) => ({
  persist: () => {},
  target: {
    name: identifier,
    value: text,
  },
});
