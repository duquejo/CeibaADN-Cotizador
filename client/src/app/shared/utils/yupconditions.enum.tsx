export enum yupConditions {
    minStringLength=2,
    maxStringLength=50
}