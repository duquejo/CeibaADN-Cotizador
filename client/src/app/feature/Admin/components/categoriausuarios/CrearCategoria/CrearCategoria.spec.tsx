import * as React from 'react';
import { RenderResult, fireEvent, render, wait } from '@testing-library/react';
import { SinonStub, stub } from 'sinon';
import { CrearCategoria } from '.';
import { setTextEvent } from 'app/shared/utils/test';

describe('CrearCategoria test', () => {
    
  let componentWrapper: RenderResult;
  let componentProps: React.ComponentProps<typeof CrearCategoria> & {
    onSubmit: SinonStub;
  };

  beforeEach(() => {
    componentProps = {
      onSubmit: stub(),
    };
    componentWrapper = render(<CrearCategoria { ...componentProps } />);
  });

  it('Debería coincidir con el snapshot', () => {
    expect( componentWrapper.container ).toMatchSnapshot();
  });

  it('Debería fallar al no tener los valores obligatorios presentes', async () => {

    // Arrange
    const elem = componentWrapper.container;
    const submitButton = elem.querySelector('button[type="submit"]');

    // Act
    await wait(() => {
      submitButton && fireEvent.click( submitButton );
    });

    // Assert
    const spans = elem.querySelectorAll('span');
    expect( spans.length ).toBe(3);
    expect( spans[0].textContent ).toBe('El campo título es requerido.');
    expect( spans[1].textContent ).toBe('El campo valor alta es requerido');
    expect( spans[2].textContent ).toBe('El campo valor baja es requerido');
  });

  it('Debería fallar con dos campos obligatorios restantes', async () => {
    
    // Arrange
    const elem = componentWrapper.container;
    const title = elem.querySelector('input[name="title"]');
    const submitButton = elem.querySelector('button[type="submit"]');

    // Act
    await wait(() => {
      title && fireEvent.change(title, setTextEvent('title', 'Categoría de prueba'));
    });
    await wait(() => {
      submitButton && fireEvent.click( submitButton );
    });    

    // Assert
    const spans = elem.querySelectorAll('span');
    expect( spans.length ).toBe(2);
    expect( spans[0].textContent ).toBe('El campo valor alta es requerido');
    expect( spans[1].textContent ).toBe('El campo valor baja es requerido');
  }); 

  it('Debería fallar con un campo obligatorio restante', async () => {
    
    // Arrange
    const elem = componentWrapper.container;
    const title = elem.querySelector('input[name="title"]');
    const priceHigh = elem.querySelector('input[name="priceHigh"]');
    const submitButton = elem.querySelector('button[type="submit"]');

    // Act
    await wait(() => {
      title && fireEvent.change(title, setTextEvent('title', 'Categoría de prueba'));
    });
    await wait(() => {
      priceHigh && fireEvent.change(priceHigh, setTextEvent('priceHigh', '50000' ));
    });
    await wait(() => {
      submitButton && fireEvent.click( submitButton );
    });    

    // Assert
    const spans = elem.querySelectorAll('span');
    expect( spans.length ).toBe(1);
    expect( spans[0].textContent ).toBe('El campo valor baja es requerido');
  }); 

  it('Debería realizar el submit con la información básica', async () => {

    // Arrange
    const elem = componentWrapper.container;
    const title = elem.querySelector('input[name="title"]');
    const priceHigh = elem.querySelector('input[name="priceHigh"]');    
    const priceLow = elem.querySelector('input[name="priceLow"]');    
    const submitButton = elem.querySelector('button[type="submit"]');

    // Act
    await wait(() => {
      title && fireEvent.change(title, setTextEvent('title', 'Calendario de prueba'));
    });
    await wait(() => {
      priceHigh && fireEvent.change(priceHigh, setTextEvent('priceHigh', '50000' ));
    });
    await wait(() => {
      priceLow && fireEvent.change(priceLow, setTextEvent('priceLow', '60000' ));
    });        
    await wait(() => {
      submitButton && fireEvent.click(submitButton);
    });

    // Assert
    const formSubmitted = componentProps.onSubmit.firstCall.args[0];
    expect( formSubmitted.nombre ).toBe('Calendario de prueba');
    expect( formSubmitted.valorAlta ).toBe(50000);
    expect( formSubmitted.valorBaja ).toBe(60000);
  });

  it('Debería realizar el submit con toda la información', async () => {

    // Arrange
    const elem = componentWrapper.container;
    const title = elem.querySelector('input[name="title"]');
    const description = elem.querySelector('textarea[name="description"]');
    const priceHigh = elem.querySelector('input[name="priceHigh"]');    
    const priceLow = elem.querySelector('input[name="priceLow"]');
    const submitButton = elem.querySelector('button[type="submit"]');

    // Act
    await wait(() => {
      title && fireEvent.change(title, setTextEvent('title', 'Calendario de prueba'));
    });
    await wait(() => {
      description && fireEvent.change(description, setTextEvent('description', 'Descripción del calendario de prueba'));
    });
    await wait(() => {
      priceHigh && fireEvent.change(priceHigh, setTextEvent('priceHigh', '50000' ));
    });
    await wait(() => {
      priceLow && fireEvent.change(priceLow, setTextEvent('priceLow', '60000' ));
    });  
    await wait(() => {
      submitButton && fireEvent.click(submitButton);
    });

    // // Assert
    const formSubmitted = componentProps.onSubmit.firstCall.args[0];
    expect( formSubmitted.nombre ).toBe('Calendario de prueba');
    expect( formSubmitted.descripcion ).toBe('Descripción del calendario de prueba');
    expect( formSubmitted.valorAlta ).toBe(50000);
    expect( formSubmitted.valorBaja ).toBe(60000);
  });
});