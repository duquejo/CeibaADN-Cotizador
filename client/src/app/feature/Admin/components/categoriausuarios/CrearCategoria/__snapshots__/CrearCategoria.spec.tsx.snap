// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CrearCategoria test Debería coincidir con el snapshot 1`] = `
<div>
  <form
    class="category__form"
    novalidate=""
  >
    <h3>
      Crear categoría de usuarios
    </h3>
    <label
      for="title"
    >
      <b>
        Título
      </b>
    </label>
    <input
      class="sc-AxjAm fjyWPa"
      id="title"
      name="title"
      placeholder="Título de la categoría, por ejemplo: 'Asociados categoría 1'"
      value=""
    />
    <label
      for="description"
    >
      <b>
        Descripción
      </b>
    </label>
    <textarea
      class="sc-AxiKw gNuUer"
      id="description"
      name="description"
      placeholder="Descripción de la categoría, por ejemplo: 'Categoría para asociados seccional Medellín'"
      rows="3"
    />
    <label
      for="price_high"
    >
      <b>
        Valor temporada alta
      </b>
    </label>
    <input
      class="sc-AxjAm fjyWPa"
      id="priceHigh"
      min="0"
      name="priceHigh"
      placeholder="Valor para días de temporada alta, por ejemplo: 50000, sin espacios y sin caracteres especiales"
      type="number"
      value=""
    />
    <label
      for="price_low"
    >
      <b>
        Valor temporada baja
      </b>
    </label>
    <input
      class="sc-AxjAm fjyWPa"
      id="priceLow"
      min="0"
      name="priceLow"
      placeholder="Valor para días de temporada baja, por ejemplo: 35000, sin espacios y sin caracteres especiales"
      type="number"
      value=""
    />
    <button
      class="sc-AxirZ iHGsua"
      type="submit"
    >
      Crear categoría de usuarios
    </button>
  </form>
</div>
`;
